### PR TITLE
Switch to Cardano haskell package repository

### PR DESCRIPTION
**Pull request by andreabedini**
@michaelpj and I have been working to establish an ~internal~ custom haskell package repository. This simplifes cabal's configuration and improves DX.

The "repository" is a static website that implements the same paths as hackage-server and it is hosted on [GitHub](https://github.com/input-output-hk/cardano-haskell-packages). The repository is built from a static description of its content, with the idea that teams will be able to publish a package with a simple PR, see the repository's [README](https://github.com/input-output-hk/cardano-haskell-packages/blob/main/README.md) for details.

This patch:

- Removes all source-repository-package stanzas (package repository has
  been pre-populated with the exact same versions)
- Removes cabal work-around disabling tests for packages erroneously
  considered local. Cabal now does the right thing
- Removes cabalWrapped, as now unnecessary

### CODE REVIEW

File: cabal.project
Comment (andreabedini) [inline cabal.project:69]: Do we need this? The [documentation](https://cabal.readthedocs.io/en/stable/cabal-project.html#cfg-field-tests) seems to suggest it is not useful:

> Force test suites to be enabled. For most users this should not be needed, as we always attempt to solve for test suite dependencies, even when this value is False; furthermore, test suites are automatically enabled if they are requested as a built target.

Comment (michaelpj) [reply cabal.project:69]: Hmm, I think in the past it used to be necessary so cabal wouldn't re-solve for tests when you switched to running tests, perhaps they fixed that. But that's an orthogonal issue I guess.

File: cabal.project
Comment (michaelpj) [inline cabal.project:1]: Comment?

File: cabal.project
Comment (michaelpj) [inline cabal.project:12]: update comment to explain how to update the other one. Or maybe a combined comment about both

File: cabal.project
Comment (michaelpj) [inline cabal.project:162]: This one should remain a `source-repository-package`, it's a fork of a Hackage package, and I'm not sure it meets our criteria for being patched in the repo.

File: cabal.project
Comment (michaelpj) [inline cabal.project:320]: also this one

File: cabal.project
Comment (michaelpj) [reply cabal.project:162]: We should also remove it from the repo.

File: nix/haskell.nix
Comment (michaelpj) [inline nix/haskell.nix:55]: also in `nix/workbench/shell.nix`

File: nix/haskell.nix
Comment (michaelpj) [inline nix/haskell.nix:54]: out of the diff above, but I suggest we also take this opportunity to remove all the `packages` arguments to `shellFor`, they shouldn't be needed

File: cabal.project
Comment (andreabedini) [reply cabal.project:162]: :+1:

File: nix/haskell.nix
Comment (andreabedini) [reply nix/haskell.nix:55]: :+1:

File: cabal.project
Comment (andreabedini) [reply cabal.project:320]: :+1:

File: cabal.project
Comment (andreabedini) [reply cabal.project:13]: I repeated the comment (adapted)

File: nix/haskell.nix
Comment (andreabedini) [reply nix/haskell.nix:52]: Sorted

File: cabal.project
Comment (andreabedini) [reply cabal.project:5]: I think we won't rely on this very much. Removed for now and using its default value (also from the repo's README).

File: cabal.project
Comment (andreabedini) [reply cabal.project:2]: :+1:

File: cabal.project
Comment (andreabedini) [reply cabal.project:72]: yes, before you had to do `tests: False` and enabled them one by one for the local packages. This is now solved (which is why I left `tests: True` which was the original intent). Now I wonder if one needs to enable them at all.

**Comment by michaelpj** [comment]
@kosyrevSerge there is some code in the workbench scripts that scrapes the `cabal.project` for dependency versions and is now failing:
- https://github.com/input-output-hk/cardano-node/blob/master/nix/workbench/run.sh#L691
- https://github.com/input-output-hk/cardano-node/blob/master/nix/workbench/manifest.sh#L117

We *think* that can just be replaced by using `ghc-pkg field --simple-output <package> version`, which will give the package version. That's what I think we want now that we're fetching released versions from a repository instead. Does that sound right to you?

**Comment by erikd** [comment]
From the sound of the "Internal Hackage repository" its *inernal" and not accessible by people outside IOG. If that is the case, then how are people outside IOG supposed to build our software?

File: cabal.project
Comment (ltouro) [inline cabal.project:1]:
```suggestion
-- Custom repository for cardano haskell packages, see CONTRIBUTING for more
```

Comment (ltouro) [reply cabal.project:1]: Foliage describe itself as a tool to create custom or private Haskell package repositories. Suggest to use the word 'custom' instead of 'internal' per erikd remarks as the repo is indeed public (https://input-output-hk.github.io/cardano-haskell-package-repo/)

**Comment by erikd** [comment]
Even if it is publicly available, we need to look at at how much more difficult (if at all) the new way of accessing packages is in comparison to the old method.

**Comment by michaelpj** [comment]
> From the sound of this "Internal Hackage repository" its internal and not accessible by people outside IOG. If that is the case, then how are people outside IOG supposed to build our software?

It's not. It's publicly accessible. There should be no difference for anyone building our software, this is all understood perfectly well by cabal. The most you need to do is call `cabal update`.

Let's try and avoid the "internal" descriptor. I've been trying to call it just the "Cardano Haskell package repository", which is a bit of a mouthful but hopefully not misleading.

**Comment by andreabedini** [comment]
In https://github.com/input-output-hk/cardano-node/pull/4411/commits/bf585f51ab77cd953bbaba417899fa9ac4cc1a61 I have temporarily disabled parts of the workbench scripts. The script extracted the commit has of some cardano components from the `cabal.project` file, information which is not available anymore. Probably the best equivalent would be either the versions from the install plan (which AFAIU is not available when those scripts run) or just the repos and their index-state. It kinda depends on what the goal of those scripts is.

**Comment by ch1bo** [comment]
> "Cardano Haskell package repository", which is a bit of a mouthful but hopefully not misleading.

We could call it `C`ardano `H`askell `P`ackage re`P`ositor`Y` -> `CHAPPY` :dog: .. unfortunately spelled wrong to be a reference to [this movie](https://en.wikipedia.org/wiki/Chappie_(film))

**Comment by andreabedini** [comment]
Another options is to just drop the "repository": Cardano Haskell Packages. "Repository" is a confusing word here because it has two meaning (as in Hacakge and in git).

**Comment by andreabedini** [comment]
Note to self, I made some cabal file revisions on the repository that end up breaking the build plan for cardano-node :disappointed: Do not bump the index-state until I fix the issue. :pray:

**Comment by andreabedini** [comment]
> Note to self, I made some cabal file revisions on the repository that end up breaking the build plan for cardano-node disappointed Do not bump the index-state until I fix the issue. pray

We run into a bit of a roadblock with haskell.nix not liking cabal file revisions. As an emergency measure we removed them from the repository, so we can unblock this PR.

**Comment by ch1bo** [comment]
> haskell.nix not liking cabal file revisions

Good haskell.nix. Allowing released code to be revised was never a good idea also IMO.

**Comment by andreabedini** [comment]
> Good haskell.nix. Allowing released code to be revised was never a good idea also IMO.

cabal file revisions don't change the code, only the metadata about it so we know what version works with what other version. We can do it in a completely declarative and reproducible way so this should not be a problem.

**Comment by andreabedini** [comment]
@erikd 

> Even if it is publicly available, we need to look at at how much more difficult (if at all) the new way of accessing packages is in comparison to the old method.

See [this repo](https://github.com/andreabedini/test-plutus-tx) for an example downstream consumer. It should work with just `cabal build`. Note that I have `libsecp256k1` installed in my system, I don't remember it is needed or not.

File: nix/workbench/manifest.sh
Comment (deepfire) [inline nix/workbench/manifest.sh:119]: Knowing exact gitrevs for key dependencies is useful, since it simplifies discussions around reproduction.

Can we make the `cardanoHaskellPackages` (and I specifically mean the Nix flake) supply the gitrevs for all constituent packages it provides?

Then we could plug in this information here.

Comment (abailly-iohk) [reply nix/workbench/manifest.sh:119]: I maybe missing some context but if you depend on versions, is not even better? The versions are not supposed to be mutable either.

File: nix/workbench/run.sh
Comment (deepfire) [inline nix/workbench/run.sh:711]: Can we avoid this (seemingly) unnecessary change?

Comment (abailly-iohk) [reply nix/workbench/run.sh:711]: What are the formatting rules for bash scripts in this repository?

File: nix/workbench/shell.nix
Comment (deepfire) [inline nix/workbench/shell.nix:90]: Ok, this, I think, is the biggest problem I can see in this PR -- this very specifically breaks Cabal-inside-Nix-shell.

I think we agreed in the past we shouldn't break it.

File: nix/workbench/manifest.sh
Comment (michaelpj) [reply nix/workbench/manifest.sh:119]: I think it would be more appropriate to use the package versions? That should be enough for reproducibility, right? I think @andreabedini had a way to do that with just `ghc-pkg`, so that should be straightforward.

**Review by deepfire — CHANGES_REQUESTED** [review]

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: In general, this is going to be at git revision granularity -- since we don't always work with released versions of packages.

Think feature branch development, for example.

So, having gitrevs uniformly available would be a consistent approach.

Having versions _in addition_ -- would be nice as well.

File: nix/workbench/shell.nix
Comment (michaelpj) [reply nix/workbench/shell.nix:90]: Can you be specific about what you think is broken? I think it all works.

File: nix/workbench/manifest.sh
Comment (andreabedini) [reply nix/workbench/manifest.sh:119]: I tried to update this part but I couldn't really do it without understanding its intended use. I don't think there's a way to push the gitrevs through the index. I suggest we stick to versions. @deepfire, just to make ure I understand, what are gitrevs useful for? looking/finding the right source used? We are thinking of adding html pages for the index, they could include a link to the corresponding source (maybe).

Comment (andreabedini) [reply nix/workbench/manifest.sh:119]: > In general, this is going to be at git revision granularity -- since we don't always work with released versions of packages.

right, well, perhaps we can change this script to list both the gitrevs from `source-package-repository` stanzas in `cabal.project` and normal dependencies from the plan. if you add a `source-package-repository` for some in work in progress depedency, they will override the depedency in the repository and show up here.

What do you think?

File: nix/workbench/shell.nix
Comment (andreabedini) [reply nix/workbench/shell.nix:90]: cabalWrapped fixes cabal's behaviour of always recompiling source-repository-packages, this does not happen anymore since dependencies are now (mostly) from a repository.

Comment (deepfire) [reply nix/workbench/shell.nix:90]: @michaelpj, well:

```

Comment (deepfire) [reply nix/workbench/shell.nix:90]: The `cardano-node` `master` branch has this bit of functionality which explains the basic workbench workflow: https://github.com/input-output-hk/cardano-node#workbench-a-local-cluster-playground

Comment (michaelpj) [reply nix/workbench/shell.nix:90]: Oh I see, we just need to provide `cabal` instead. I just assumed it was there!

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: > I don't think there's a way to push the gitrevs through the index.

@angerman, do you think we have no options here?

Comment (michaelpj) [reply nix/workbench/manifest.sh:119]: Why do you need the gitrevs for _released_ package versions? Getting them when you're using a `source-repository-package` seems essential indeed, but if you are using a package from the repo, surely the version is sufficient?

File: nix/workbench/shell.nix
Comment (deepfire) [reply nix/workbench/shell.nix:90]: If both `cabal` and `cabalWrapped` were there, there'd be a clash, because both of them provide the same executable name.

Comment (deepfire) [reply nix/workbench/shell.nix:90]: @michaelpj, but a question -- what is the motivation for `cabalWrapped` removal, in the first place?

Can this PR be made to work with `cabalWrapped` still there?  How difficult would that be?

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: @michaelpj, again, this is just simpler when you deal with reproductions and general discussion of changes.

When you have the gitrev available, that's a hard source of truth.  Whereas with git tags it's sometimes not exactly a 1-1 relationship with cabal package versions, since it's managed by humans.

File: nix/workbench/shell.nix
Comment (michaelpj) [reply nix/workbench/shell.nix:90]: It's not needed, because `cabal` just works - like it will for every other project.

Comment (andreabedini) [reply nix/workbench/shell.nix:90]: ok, we can fix this, thanks for letting us know

Comment (michaelpj) [reply nix/workbench/shell.nix:90]: I know you have a slightly different workflow, Serge, I tried to discuss that a bit here: https://docs.google.com/document/d/1PGmrvcNmfLP7HhGZpcgEee6zlHsQ-hozCC2XkjzHcbI/edit#heading=h.q6n5jsdrrm3b

If you find you still use `source-repository-package` lots then `cabalWrapped` might still be useful for you.

Comment (deepfire) [reply nix/workbench/shell.nix:90]: @michaelpj, running node clusters with feature branches is something not specific to the performance team.

I've made a reply in the document: https://docs.google.com/document/d/1PGmrvcNmfLP7HhGZpcgEee6zlHsQ-hozCC2XkjzHcbI/edit?disco=AAAAf9tvVhs

In fact, we specifically want to encourage such use -- precisely to enable people to get direct, early feedback on their code changes.

Comment (deepfire) [reply nix/workbench/shell.nix:90]: In short, I don't think we have evidence that `cabalWrapped` is necessary in this (otherwise great!) PR.

I propose we undo that change and reinstate it.

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: @andreabedini, so from what @angerman says, the information is available somewhere -- can we make it available in the flake as well?

Comment (angerman) [reply nix/workbench/manifest.sh:119]: @deepfire its in the toml file in the cardano-Haskell-packages repo.

You need to pin the repo through a nix input anyway, and we could then just read the rev from the toml file in nix from there for the package/version; if needed.

File: nix/workbench/shell.nix
Comment (andreabedini) [reply nix/workbench/shell.nix]: I put `cabalWrapped` back. Thanks @deepfire.

File: nix/workbench/run.sh
Comment (andreabedini) [reply nix/workbench/run.sh]: I reverted this change.

File: nix/workbench/manifest.sh
Comment (michaelpj) [reply nix/workbench/manifest.sh:119]: > again, this is just simpler when you deal with reproductions and general discussion of changes.

I don't agree with this. Revisions are long, confusing, don't obviously have an ordering. Versions are much easier to deal with 90% of the time. If you're dealing with something unreleased, sure you need revisions, but I still really don't see how a discussion about released versions could be ambiguous when using versions.

**Comment by deepfire** [comment]
Thank you @michaelpj @andreabedini !

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: I never said versions are bad to have!

Only suggesting that losing the gitrevs is unfortunate, because gitrevs are precise and not subject to human errors and other deviations.

Having a cabal version, on the other hand, does not guarantee you a gitrev.

Comment (angerman) [reply nix/workbench/manifest.sh:119]: @deepfire while it is true that _hackage.haskell.org_ packages do not have revisions, due to the design of hackage allowing arbitrary source management, and a such just release tarball uploads, `cardano-haskell-packages` _does_ have revisions. So at least for `cardano-haskell-packages` you can go from version number to git revision e.g. [measures-0.1.0.0](https://raw.githubusercontent.com/input-output-hk/cardano-haskell-packages/main/_sources/measures/0.1.0.0/meta.toml);

@andreabedini maybe we should ensure that people don't put _tags_ as revisions in the url, as those may change over time (or resolve them at the point of index constructions).

Comment (andreabedini) [reply nix/workbench/manifest.sh:119]: @angerman At the moment we are pointing to arbitrary tarballs, when I implement a proprer github backend I can do this.

File: cabal.project
Comment (ch1bo) [inline cabal.project:332]: :heart_eyes:

File: cabal.project
Comment (shlevy) [inline cabal.project:99]: Might be nice to have documentation saying when downstream repos need these overrides as well.

Comment (andreabedini) [reply cabal.project:99]: Oh dear yes.

**Review by deepfire — APPROVED** [review]

File: cabal.project
Comment (michaelpj) [reply cabal.project:99]: Happy to add this, just not sure where. Probably this should be in the README. I'll add a section about "using the Node as a Haskell dependency".

File: nix/workbench/manifest.sh
Comment (deepfire) [reply nix/workbench/manifest.sh:119]: @andreabedini, in the interim it'd be nice to have `cardanoHaskellPackages` provide at least the versions in some kind of a `package-name -> version string` attrset.

**Comment by michaelpj** [comment]
I think this is ready to go. I added a note about using the `cardano-node` packages from elsewhere, and consolidated some development documentation.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
I'm generally for this but I will take a closer look after the cli release

File: cabal.project
Comment (ch1bo) [reply cabal.project:99]: Could we get that info on haskell package level? i.e. likely these won't be needed if we only use `cardano-api`?

Comment (michaelpj) [reply cabal.project:99]: I don't know the answer, unfortunately.

**Comment by andreabedini** [comment]
@michaelpj I suggest a stronger wording (and a clear policy)

```
--
-- Open PR upstream, maintainer unresponsive, hopefully short-lived fork.
--

-- Note: these dependencies are at project level and not at package level.
-- This means that they won't be part of the dependencies of any package
-- released from this repository and they won't be seen by downstream
-- consumers.
-- It is ok to keep these during development but it is essential that
-- they are not needed by any released package.
```

What do you think?

**Comment by michaelpj** [comment]
I'm not sure if that's realistic. Maybe if we just release a patched version of `ekg-json` to CHaP using our padding scheme we can get past it. I'm unsure about `threepenny-gui`. It would be nice to enforce a "no `source-repository-package`s at release time policy, though.

**Comment by andreabedini** [comment]
> It would be nice to enforce a "no `source-repository-package`s at release time policy, though.

:crossed_fingers:

**Comment by michaelpj** [comment]
@Jimbo4350 if the release is going ahead off the branch, can we move forward with this on master?

File: CONTRIBUTING.rst
Comment (Jimbo4350) [inline CONTRIBUTING.rst:135]: I don't follow. Checkout the repository with the dependencies "beside" the `cardano-node` checkout?

File: CONTRIBUTING.rst
Comment (Jimbo4350) [inline CONTRIBUTING.rst:145]: This seems unnecessarily complex to simply bump dependencies in the node in order to try them out. Is there a way to generate a `cabal.project` file from CHaP? I could then just use that locally until I'm satisfied the dep bumps build successfully. I do not want to use Nix.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
A few comments

File: CONTRIBUTING.rst
Comment (michaelpj) [reply CONTRIBUTING.rst:5]: This isn't new: I just moved it from the README. I'll add a link though.

File: CONTRIBUTING.rst
Comment (michaelpj) [reply CONTRIBUTING.rst:135]: old content

File: CONTRIBUTING.rst
Comment (michaelpj) [reply CONTRIBUTING.rst:145]: old content. using s-r-p works, as noted outside this big workbench section

**Comment by michaelpj** [comment]
@Jimbo4350 squashed and hydra passed.

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cabal.project
Comment (erikd) [reply cabal.project:71]: When using `source-repository-package` the directive `tests: True` turns on tests within all `source-repository-package` dependencies, which (especially higher up the stack) just needless runs tests in dependencies for which it should be possible to avoid re-rumning tests. 

In say `db-sync` we have `tests: False` to disable all tests and the specific stanzas to turn on tests for packages in the `db-sync` repo.

Comment (michaelpj) [reply cabal.project:71]: > When using source-repository-package the directive tests: True turns on tests within all source-repository-package dependencies, which (especially higher up the stack) just needless runs tests in dependencies for which it should be possible to avoid re-rumning tests.

Is this still true in recent versions of cabal? We have `tests: True` in `plutus` and I've never observed it running or building tests for s-r-ps, and I tried again just now.

**Pull request merged**
